### ISSUE TITLE
DualSense Audio Support

### ISSUE THREAD

**Issue by citrizon** [bluez/bluez#892]
Hello! I am an Arch Linux user and i'm powering my bluetooth stack using Bluez. I configured my system and everything works fine.
I can connect my DS4 controller to my PC, however when i try to connect a headphone, there is no sound output. I am checking the error through `dmesg`. the result is here:
```r
[ 3120.827352] input: Wireless Controller as /devices/virtual/misc/uhid/XXXX:XXXX:XXXX:XXXX/input/input34
[ 3120.828103] input: Wireless Controller Motion Sensors as /devices/virtual/misc/uhid/XXXX:XXXX:XXXX:XXXX/input/input35
[ 3120.828493] input: Wireless Controller Touchpad as /devices/virtual/misc/uhid/XXXX:XXXX:XXXX:XXXX/input/input36
[ 3120.829440] playstation XXXX:XXXX:XXXX:XXXX: Registered DualShock4 controller hw_version=0x00006404 fw_version=0x00007008
[ 3174.454748] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=24
[ 3174.465644] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
[ 3174.482671] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
[ 3174.501570] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
[ 3174.512648] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
[ 3174.536696] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
[ 3174.549577] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
[ 3174.561715] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
[ 3174.579633] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
[ 3174.596593] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
```
(Potentially sensitive information is redacted.)

I would like to learn about this problem and i would like to contribute to fix this issue (hopefully... maybe)

**Comment by Vudentz** [comment]
> Hello! I am an Arch Linux user and i'm powering my bluetooth stack using Bluez. I configured my system and everything works fine. I can connect my DS4 controller to my PC, however when i try to connect a headphone, there is no sound output. I am checking the error through `dmesg`. the result is here:
> 
> ```r
> [ 3120.827352] input: Wireless Controller as /devices/virtual/misc/uhid/XXXX:XXXX:XXXX:XXXX/input/input34
> [ 3120.828103] input: Wireless Controller Motion Sensors as /devices/virtual/misc/uhid/XXXX:XXXX:XXXX:XXXX/input/input35
> [ 3120.828493] input: Wireless Controller Touchpad as /devices/virtual/misc/uhid/XXXX:XXXX:XXXX:XXXX/input/input36
> [ 3120.829440] playstation XXXX:XXXX:XXXX:XXXX: Registered DualShock4 controller hw_version=0x00006404 fw_version=0x00007008
> [ 3174.454748] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=24
> [ 3174.465644] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
> [ 3174.482671] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
> [ 3174.501570] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
> [ 3174.512648] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
> [ 3174.536696] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
> [ 3174.549577] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
> [ 3174.561715] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
> [ 3174.579633] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
> [ 3174.596593] playstation XXXX:XXXX:XXXX:XXXX: Unhandled reportID=19
> ```
> 
> (Potentially sensitive information is redacted.)
> 
> I would like to learn about this problem and i would like to contribute to fix this issue (hopefully... maybe)

Doesn't seem like it is using A2DP/HFP for audio, so I guess it is transmitted over HID so it needs a kernel driver that can handle these reports.

**Comment by mirh** [comment]
You are correct. Previous episodes:
https://github.com/chrippa/ds4drv/pull/80
https://github.com/sensepost/dual-pod-shock/issues/1

**Comment by egormanga** [comment]
Recent discoveries in this field: https://github.com/egormanga/SAxense (by me). I tend to think this is possible to implement mostly (or completely) outside BlueZ, i.e. in PipeWire.

**Comment by mirh** [comment]
Possible? Sure. Even because I guess the microphone eventually needs opus somewhere (and probably there's quite some things that should also go into hid-playstation). But is it the best place for everything?

In particular I was kinda worried about the "legitimacy" of the vanilla quadraphonic setup. If you have such configuration the sound that plays in the front speakers isn't the same as if you were stereo (let alone that obviously the system has no clue it's even headphones).
So I feel like the thing should be exposed to the system as such, and then something in the stack takes care of mirroring them to channel 3 and 4 (or you properly go through HAPTICS_SELECT bit to send rumble with hid, but then something still has to take care of the whole process).

**Comment by egormanga** [comment]
No, it's only quad-channel when wired. Over BT it's two separate stereo streams, and you're right to guess that the non-haptic ones are Opus.

**Comment by mirh** [comment]
I read that you are tinkering with pipewire and bluez.. 
Maybe you could have a dedicated RFC on lkml or chime in here (guy is presumably even fresh of digging some datasheets)
https://lore.kernel.org/lkml/CAEc3jaAm4gws8drVKvQGYHoAT4fb5i7sSPJ0LFiSGHaVAzKZnA@mail.gmail.com/
https://lore.kernel.org/lkml/CAEc3jaBkyPgnWjkZX6XjT6YKn4sAmHpOrtwp_+UMruS-hwUz_Q@mail.gmail.com/T/#u
